### PR TITLE
unify predict functionality

### PR DESCRIPTION
**Pull request by felixbur**
as discussed in [this issue](https://github.com/felixbur/nkululeko/issues/369)

* Replaced nkululeko.demo, nkululeko.feature_demo, and nkululeko.testing with an expanded nkululeko.predict supporting --file/--folder/--list/--mic and --type feats|model.
* Added extensive unit tests for the new predict helpers/CLI/dispatch behavior and added SNR estimator regressions for edge cases (silence, constant signal, short signal).
* Updated docs, scripts, and packaging metadata to reflect the unified module and removed deprecated entry points/modules.

take as command line argument

    --file: argument to predict label from AUDO FILE(s), can be a single audio file or several audio files, output to FILE_result.txt.
    --folder: argument to receive input in the form of a folder (auto-detecting audio files)
    --list: argument to predict label from CSV FILE containing audio path. If this list already has columns, keep them! and keep the index intact!
    --outfile: (optional) argument to save predicted labels from given list, defaults to ./prediction_result.csv
    --mic: argument to work directly on MIC input and output to stdout
    --model: name of prediction model, e.g. age, gender, mos, ... (must be implemented in nkululeko)
    --type: either feats or model, if feats; the predictor is described in the FEATS section, else the experiment is loaded from --config and the best model is being used (specified in the MODEL section) (defaults to feats)
    --config ini.file with values for all the above, if they are not overwritten on the command line.

### CODE REVIEW

**Comment by bagustris** [comment]
@felixbur 

Could you remark on all issues by Copilot above first (if already revised)?
I'll test them locally.

File: nkululeko/feat_extract/feats_audwav2vec2.py [hunk 1]
```diff
@@ -74,6 +74,10 @@ def extract(self):
                         )
                     f_df = self.extract_sample_df(signal, sr, (file, start, end))
                 except Exception as ror:
+                    tmp_file_name = "tmp_audio.wav"
+                    if signal is not None and sr is not None:
+                        audiofile.write(tmp_file_name, signal, sr)
+                        self.util.debug(f"segment from {file} written to {tmp_file_name} for debugging.")
```
[the remark this answers is not in the document]
Comment (felixbur) [reply nkululeko/feat_extract/feats_audwav2vec2.py]: changed

File: docs/source/text_processing.md [hunk 2]
```diff
@@ -1,285 +1,206 @@
...
 ### Output
 
-The output CSV will include a new `translation` column:
-
 ```csv
-file,emotion,text,translation
-./data/emodb/wav/03a01Fa.wav,happiness,Der Lappen liegt auf dem Eisschrank.,The rag is on the freezer.
+file,start,end,emotion,text,translation
+./data/emodb/wav/03a01Fa.wav,0 days,,happiness,Der Lappen liegt auf dem Eisschrank.,The rag is on the freezer.
 ```
```
[the remark this answers is not in the document]
Comment (felixbur) [reply docs/source/text_processing.md:105]: changed

File: nkululeko/predict.py [hunk 3]
```diff
@@ -1,70 +1,821 @@
...
+        path_col = in_df.columns[0]
+        files = in_df[path_col].astype(str).tolist()
+        seg_index = audformat.segmented_index(
+            files=[os.path.abspath(f) for f in files],
+            starts=[pd.Timedelta(0)] * len(files),
+            ends=[pd.NaT] * len(files),
+        )
+        seg_index = _resolve_nat_ends(seg_index)
+        seg_df = in_df.drop(columns=[path_col]).copy()
+        seg_df.index = seg_index
```
[the remark this answers is not in the document]
Comment (felixbur) [reply nkululeko/predict.py:458]: no, i think this makes it too complicated, lists should always have file as the first column

File: nkululeko/predict.py [hunk 4]
```diff
@@ -1,70 +1,821 @@
...
+    if args.mic:
+        _run_mic(args, util)
+    elif args.file:
+        _run_files(args.file, args, util)
+    elif args.folder:
+        _run_folder(args.folder, args, util)
+    elif args.list_path:
+        _run_list(args.list_path, args, util)
+    else:
+        util.error("no input given: provide one of --file, --folder, --list, --mic")
```
[the remark this answers is not in the document]
Comment (felixbur) [reply nkululeko/predict.py]: yes, changed

File: nkululeko/predict.py [hunk 5]
```diff
@@ -1,70 +1,909 @@
...
+        if "databases" not in config["DATA"]:
+            config["DATA"]["databases"] = "['adhoc']"
+        return config
+
+    config = configparser.ConfigParser()
+    tmp_root = tempfile.mkdtemp(prefix="nkulu_predict_")
+    config["EXP"] = {"root": tmp_root, "name": "predict"}
+    config["DATA"] = {"databases": "['adhoc']", "target": "predicted"}
+    config["FEATS"] = {"no_reuse": "True", "needs_feature_extraction": "True"}
+    config["MODEL"] = {}
```
[the remark this answers is not in the document]
Comment (felixbur) [reply nkululeko/predict.py:283]: done

File: nkululeko/feat_extract/feats_audwav2vec2.py [hunk 6]
```diff
@@ -74,6 +76,10 @@ def extract(self):
                         )
                     f_df = self.extract_sample_df(signal, sr, (file, start, end))
                 except Exception as ror:
+                    tmp_file_name = f"tmp_{uuid.uuid4()}.wav"
+                    if signal is not None and sr is not None:
+                        audiofile.write(tmp_file_name, signal, sr)
+                        self.util.debug(f"segment from {file} written to {tmp_file_name} for debugging.")
                     self.util.error(f"error {ror} on file {file}")
```
[the remark this answers is not in the document]
Comment (felixbur) [reply nkululeko/feat_extract/feats_audwav2vec2.py:86]: it writes to current dir and this should be writeable in any case

**Comment by felixbur** [comment]
> @felixbur
> 
> Could you remark on all issues by Copilot above first (if already revised)? I'll test them locally.

done

**Comment by bagustris** [comment]
@felixbur 

There are two new comments from Copilot with High and Medium severity.
And here are my comments:

1. Code duplication — index extraction duplicated verbatim in two functions
                                                                                
  File: nkululeko/predict.py — _predict_with_features (lines 653–673) and       
  _predict_with_model (lines 747–767)                                           
                                                                                
  The 20-line block that extracts file, offset, duration from a segmented-index 
  tuple is copy-pasted identically in both functions. A single helper removes
  the maintenance burden:                                                       
  ```                                                                            
  def _unpack_index(idx):                      
      """Return (file, offset_s, duration_s_or_None) from a segmented index     
  entry."""                                                                     
      if isinstance(idx, tuple):                                                
          file, start, end = idx                                                
          offset = start.total_seconds() if hasattr(start, "total_seconds") else
   float(start)                                                                 
          duration = None
          if end is not pd.NaT and end is not None and not pd.isna(end):        
              dur = (end - start).total_seconds() if hasattr(end - start,       
  "total_seconds") else None                                                    
              if dur is not None and dur > 0:                                   
                  duration = dur                                                
      else:                                                                     
          file, offset, duration = idx, 0, None
      return file, offset, duration                                             
   ```                                                                           
  Then both loops become simply:         
                                       
    file, offset, duration = _unpack_index(idx)
                                                                                
  2. atexit.unregister in test may silently fail on some Python versions      
                                                                                
  File: tests/test_predict.py, line 382                                         
  ```                                                                              
  atexit.unregister(__import__("nkululeko.predict",                             
  fromlist=["_cleanup_path"])._cleanup_path)                                    
   ```                                            
  atexit.unregister does not raise if the function was never registered (it is  
  safe in Python 3.9+). However the Copilot comment about it raising ValueError
  is incorrect for Python ≥ 3.9 — the tests pass fine. That said, the comment in
   the test is confusing: it says "If a cleanup were wrongly registered, the  
  assertion below would not catch it directly" but then calls unregister
  unconditionally. The test should either:
  - Assert that config["EXP"]["root"] does not start with the nkulu_predict_
  prefix (which would catch accidental temp-dir creation), or                   
  - Drop the atexit.unregister call entirely since it tests nothing.
                                                                                
  Suggested fix:
  ```python                                                              
  config = _load_config(args)                                                   
  assert config["EXP"]["root"] == "./"
  # No temp dir should have been created when an explicit config is supplied.   
  assert "nkulu_predict_" not in config["EXP"]["root"]                          
```
                                                                                
  3. _run_list: plain CSV assumes first column is audio path — undocumented and 
  fragile                                                                       
                                                                                
  File: nkululeko/predict.py, line 448                                          
                                                                              
      path_col = in_df.columns[0]                                                   
                                                                              
  When a user saves a CSV with df.to_csv(path) (the default), pandas writes an  
  Unnamed: 0 index column first. The code silently treats that as the audio
  path, which produces confusing file-not-found errors. The PR description      
  explicitly says "keep the index intact", but the current code does not handle
  Unnamed: 0.                                  

  Suggested fix (minimal, matching your's preference to keep the contract      
  simple):
  ```python
  # Skip a leading Unnamed index column that pandas writes by default.          
  first_col = in_df.columns[0]                                                
  if first_col == "Unnamed: 0" or first_col.startswith("Unnamed:"):             
      in_df = in_df.set_index(first_col)                                        
      path_col = in_df.columns[0]                                               
  else:                                                                         
      path_col = first_col                                                    
  ````                                                                              
  Or document in the docstring and help text that "first column must be the   
  audio path, not a pandas-written integer index."                              
  
  4. feats_audwav2vec2.py: debug write to CWD still masks original exception    
                                                                              
  File: nkululeko/feat_extract/feats_audwav2vec2.py, lines 79–83                
  
  ```python                                                                            
  except Exception as ror:                     
      tmp_file_name = f"tmp_{uuid.uuid4()}.wav"                                 
      if signal is not None and sr is not None:                                 
          audiofile.write(tmp_file_name, signal, sr)    # ← can itself raise    
          self.util.debug(...)                                                  
      self.util.error(f"error {ror} on file {file}")                            
  ````
                                                                                
  audiofile.write can fail (e.g., unsupported dtype, disk full). If it does, it 
  replaces the original exception ror and util.error(...) is never reached,     
  hiding the root cause. The fix is one extra try/except:                       
 
```python                                                                             
  except Exception as ror:                     
      tmp_file_name = f"tmp_{uuid.uuid4()}.wav"
      if signal is not None and sr is not None:
          try:
              audiofile.write(tmp_file_name, signal, sr)                        
              self.util.debug(f"segment from {file} written to {tmp_file_name}
  for debugging.")                                                              
          except Exception as write_err:                                      
              self.util.warn(f"could not write debug audio: {write_err}")       
      self.util.error(f"error {ror} on file {file}")                            
```
                                               
  5. _get_feature_extractor: function falls off the end without returning None  
                                                                              
  File: nkululeko/predict.py, line 924                                          
 
```                                                                             
      util.error(f"unknown feature extractor: {model_name}")
      # implicit return None here                                               
  ```

  util.error() calls sys.exit(1), so in practice this is unreachable — but the  
  function has no explicit return after util.error. Static type-checkers flag 
  this as str | None return type ambiguity. Add:                                
  
  ```                                                                            
      util.error(f"unknown feature extractor: {model_name}")
      return None  # unreachable; util.error() calls sys.exit(1)
                                                                                
      util.error(f"unknown feature extractor: {model_name}")
      return None  # unreachable; util.error() calls sys.exit(1)
```

  6. pyproject.toml: nkululeko.multidb and nkululeko.optim entry points removed without notice

  File: pyproject.toml

  The [project.scripts] block no longer includes nkululeko.multidb or nkululeko.optim. Both modules still exist in the codebase. This appears unintentional — users who installed a prior version and rely on these CLI shortcuts will silently lose them after upgrading.

  Suggested fix: Restore the missing entry points:
  ```
  "nkululeko.multidb" = "nkululeko.multidb:main"
  "nkululeko.optim" = "nkululeko.optim:main" 
 ```

**Comment by felixbur** [comment]
thanks @bagustris , all done

File: tests/test_predict.py [hunk 7]
```diff
@@ -1,85 +1,1273 @@
...
+        cfg_path.write_text("[EXP]\nname = x\nroot = ./\n")
+
+        args = argparse.Namespace(config=str(cfg_path))
+        # If a cleanup were wrongly registered, the assertion below would
+        # not catch it directly — but the EXP.root won't point at a temp
+        # dir, which is the observable contract.
+        config = _load_config(args)
+        assert config["EXP"]["root"] == "./"
+        atexit.unregister(__import__("nkululeko.predict", fromlist=["_cleanup_path"])._cleanup_path)
+
```
[the remark this answers is not in the document]
Comment (felixbur) [reply tests/test_predict.py:382]: done

File: nkululeko/feat_extract/feats_audwav2vec2.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (felixbur) [reply nkululeko/feat_extract/feats_audwav2vec2.py:86]: done
